Improved and multithreaded dual ISO processing, including some other minor UI updates/fixes…

fijha opened the pull request:
First, thank you very much for making this app! Especially the ability to process dual ISO files. It's so useful with these older Canon cameras. Greatly improving dynamic range. Honestly, very impressive! However, there are some issues making it unusable for video footage.

I have managed to solve the 2 biggest issues with a bonus:

1. The flickering issue

2. Artifacts in blown highlights areas affecting lower bit depth footage (e.g. 12-bit, 10-bit…). Solving this also improved highlight reconstruction, especially in cinema DNG files.

As a bonus, this makes it possible to enable multithreading, which improves performance a lot, making it even more usable.

I added manual control over 3 most crucial parameters for proper exposure blending: horizontal pattern, exposure correction, and black delta. Manual tweaking is necessary in most cases.

All these parameters were calculated with every single frame, which doesn't make sense and was the root cause of the flickering issue. Now they are calculated and set only once per file. Freeing up CPU a bit.

Next there is an option for ISO-based or histogram-based automatic correction. ISO-based correction provides a good starting point on the initial file load if dual ISO metadata are present in a file. Histogram-based correction requires you to select a scene/frame with a wide (ideally full) dynamic range for accurate results.

I'm very happy with the update. Works very well for me. Hopefully not just for my setup. Tested on a mac, with lots of footage from my EOS 650D using a crop-mood from Bilal Fakhouri.

It also includes several minor UI updates/fixes not specific to dual ISO.

So enjoy! Build, test, and report :)

fijha committed:
Improved dual ISO processing, including some other minor updates/fixes

This update focuses primarily on solving 2 of the biggest dual ISO issues: the flickering issue and artifacts in blown highlights areas affecting lower bit depth footage (e.g. 12-bit, 10-bit…).

Added manual control over 3 most crucial parameters for proper exposure blending: horizontal pattern, exposure correction, and black delta.

Big performance boost by enabling multithreading.

It also includes several minor UI updates/fixes not specific to dual ISO.

masc4ii commented:
Wow. What a monster PR. I am sure you must have spent many many hours and days or even weeks to bring all this. A big thank you for your work!

The dualIso for current new DualISO files seems to work without problem here. Cool: many of my old omp lines work now - when I tried this made big problems - so it was commented. I should record more those new DualISO files to test more.

For other things I have some questions, as I think it breaks more than it fixes:
- Why did you remove DualISO Force mode? This is important and necessary for older DualISO files. Processing is now broken for these files (nearly all my test files don't work now).
- Why did you remove DualISO Preview mode? This was a fast easy preview mode (working real time), for just quick checking of files without full processing.
- Why did you remove 0.1 precision of black level? This was important for lower bit RAW files, when the correct black level was in between full numbers. We now don't get a clean black, but it is always a bit green or red.
- Did you insert focus/bad pixel fix for DualISO? Cool! Unfortunately I somehow can't test, as for my build the DownloadManager makes the app crashing... not your fault I guess.
- I don't understand the difference in "-" in InfoDialog.ui
- Double spaces in InfoDialog were done for better readability of the multiple numbers in one cell

This was no "full" test (this should work by far longer), but a first little view on all your hard work!

fijha commented:
Definitely months :D… figuring out how all of this works… I learned a lot!

Dual ISO force mode is still there. No need for another button. If you load a file without dual ISO metadata, the "On" button changes to "Force On". You can basically process any file as dual ISO file. Try resetting a receipt.

Preview mode – probably no longer necessary. Even as a fast preview, it never worked for me, and it always produced a horrible image. Actually, if you switch interpolation from "AMaZE" to "Mean", it is now almost as fast as preview mode.

The 0.1 black level precision in the current app release – looks like it only works if dual ISO is switched off. I missed that somehow… I can definitely bring that back. Or maybe there is a better way?

The update includes a temporary solution for focus/bad pixel interpolation producing red/green pixels around high-contrast edges. I was able to solve it by modifying the `interpolate_horizontal` function, which goes before dual ISO processing, but also by applying the `interpolate_rewind` method after dual ISO processing. Not ideal, but it works. Could be improved by rewriting the `interpolate_horizontal` function. But I think the reason why it sometimes produces red/green pixels around high-contrast edges is clipping in the bright exposure (pink highlights). **So here is my idea – I think implementing a highlight reconstruction on bright exposure could solve not just this but also improve overall exposure blending and details around high-contrast edges.**

Info dialog – just for consistency and typography reasons I replaced hyphens with en dashes. Looks better :). Also removed double spaces. In terms of readability, there must be a better way, maybe applying some formatting.

Thank you for the quick response and testing!

masc4ii commented:
> Dual ISO force mode is still there. No need for another button. If you load a file without dual ISO metadata, the "On" button changes to "Force On". You can basically process any file as dual ISO file. Try resetting a receipt.

Hm... I was hoping for that, but it doesn't work. I uploaded two files. The first represents >90% of the results with my test files. Nothing happens when clicking "On".

[File1](https://www.dropbox.com/scl/fi/sec6ds4i7csucp7r7kphc/14bit_lossless_dualiso.MLV?rlkey=vmonkzpdd3enupedjxvfkymqh&st=3odkauhn&dl=0)

And here the pictures becomes pink, while all lines stay in the image.

[File2](https://www.dropbox.com/scl/fi/x4ht3f4xryf1ewui6pofz/5DII-DualIso.MLV?rlkey=fxfg289v7kjgbd42pws94mwpt&st=qk02vigj&dl=0)

> Preview mode – probably no longer necessary. Even as a fast preview, it never worked for me, and it always produced a horrible image. Actually, if you switch interpolation from "AMaZE" to "Mean", it is now almost as fast as preview mode.

So I should play more! Good to know.

> The 0.1 black level precision in the current app release – looks like it only works if dual ISO is switched off. I missed that somehow… I can definitely bring that back. Or maybe there is a better way? 

Yes, please let this in the app. We need it. It was the most easy solution we found years ago. Do you have a better idea? I am not sure if it works with DualISO too... I barely used DualISO because the cost was higher then the outcome (for me).

> So here is my idea – I think implementing a highlight reconstruction on bright exposure could solve not just this but also improve overall exposure blending and details around high-contrast edges.

You mean for each kind of clip, even non-DualISO? This would be awesome. The existing solution is not great and I don't get the librtprocess algorithm to life in our app.

Thanks again for all the work!

fijha commented:
That's strange… Here are my screenshots with the files you included.

<img width="2560" height="1440" alt="Screenshot 2025-09-25 at 20 23 27" src="https://github.com/user-attachments/assets/6e6a06ce-1118-4297-9bdf-6ea1aaa2745a" />
<img width="2560" height="1440" alt="Screenshot 2025-09-25 at 20 23 45" src="https://github.com/user-attachments/assets/94dfa9e3-41dd-43cb-ade3-c8649c7b5fc6" />

Maybe a little bug with auto correction on first import… But you can always click on "Histogram" or "ISO" if available, to force it. It doesn't act just like a switch but also as a button.

masc4ii commented:
Did you change something? Here I even don't see "Force On".
<img width="1552" height="937" alt="Bildschirmfoto 2025-09-25 um 20 48 34" src="https://github.com/user-attachments/assets/51dabb00-97e6-4d67-9623-4b0067a77a80" />

I can click on whatever button and absolutely nothing happens to the image.

fijha commented:
I didn't change anything form the last commit. Can be something with your build? Cache related… or maybe QT? I'm using QT 6.9.0.

masc4ii commented:
No chance for Qt6 here. I am on Qt5.15.10, LLVM16, ARM64. But I don't think Qt version should change the processing?!

fijha committed:
Merge branch 'ilia3101:master' into Dual-ISO

fijha committed:
Update Windows.yml

fijha committed:
Update macOS-Intel.yml

fijha committed:
Update macOS-Arm64.yml

fijha committed:
Update Linux.yml

masc4ii commented:
Build via runner worked! Yay. So I installed the windows version into a Wine bottle. Unfortunately the same thing happens as I've shown before with version coming out of my compiler.

<img width="1312" height="840" alt="Bildschirmfoto 2025-09-27 um 14 32 38" src="https://github.com/user-attachments/assets/0e61b33e-283e-4730-b8e2-b6370c630bcf" />

masc4ii commented:
For very dark clips, there is a banding problem with your version. The banding effect is getting better for for higher exposure correction. Maybe some bitdepth is not huge enough for corrected subimages?

Yours:
<img width="1552" height="937" alt="Bildschirmfoto 2025-09-27 um 17 07 42" src="https://github.com/user-attachments/assets/146109fc-b4aa-4a6f-a83f-9406c6d754a5" />

Master:
<img width="1552" height="937" alt="Bildschirmfoto 2025-09-27 um 17 08 17" src="https://github.com/user-attachments/assets/8317c75f-d8a8-44cc-8503-a93c72e54d96" />

(See also my Exposure- & Lighten Slider. I used the same for both screenshots. And yes, the clip is really very very dark.)
<img width="511" height="651" alt="Bildschirmfoto 2025-09-27 um 17 10 24" src="https://github.com/user-attachments/assets/76e791b7-6245-41ff-842b-37dc309c238f" />

fijha commented:
Yes, just a temporary fix. Needs more debugging… wondering why `receipt->dualIsoForced()` does not equal `DISO_FORCED` on your system…

I have a lot of very dark dual ISO clips, exactly like yours, without issues… Maybe because it is 13-bit? I only tested 14, 12 and 10… Can you include that file or just a few frames?

fijha committed:
Dual ISO force mode fix

masc4ii commented:
You updated all the runners. Nice! Have you seen the result of the Linux runner? 

```
../../src/mlv/llrawproc/dualiso.c:1338:13: error: not enough perfectly nested loops before ‘int’
 1338 |             int s = (is_bright[y%4] == is_bright[(y+1)%4]) ? -1 : 1;    /* points to the closest row having different exposure */
      |             ^~~
../../src/mlv/llrawproc/dualiso.c:1389:66: error: ‘s’ undeclared (first use in this function)
 1389 |                             int dy1 = edge_directions[d].ack.y * s;
      |                                                                  ^
../../src/mlv/llrawproc/dualiso.c:1389:66: note: each undeclared identifier is reported only once for each function it appears in
../../src/mlv/llrawproc/dualiso.c:1427:13: error: not enough perfectly nested loops before ‘uint32_t’
 1427 |             uint32_t* native = BRIGHT_ROW ? bright : dark;
      |             ^~~~~~~~
../../src/mlv/llrawproc/dualiso.c:1449:21: error: ‘native’ undeclared (first use in this function)
 1449 |                     native[x   + y * w] = raw_get_pixel32(x, y);
      |                     ^~~~~~
../../src/mlv/llrawproc/dualiso.c: In function ‘mean23_interpolate’:
../../src/mlv/llrawproc/dualiso.c:1500:13: error: not enough perfectly nested loops before ‘uint32_t’
 1500 |             uint32_t* native = BRIGHT_ROW ? bright : dark;
      |             ^~~~~~~~
../../src/mlv/llrawproc/dualiso.c:1542:17: error: ‘native’ undeclared (first use in this function)
 1542 |                 native[x   + y * w] = raw_get_pixel32(x, y);
      |                 ^~~~~~
```

I think the linux compiler is correct here and I wonder that works on windows and macOS. Should work without the collapse(2), what do you think? Or the "int.." lines should be moved into the next for, right?

fijha commented:
I fixed dual ISO force mode :)

Yes, updated all runners, but I have no idea what is going on with linux… To be hones I'm very new to all this C/C++ compiling stuff :D… so it is the issue with multithreading?

fijha committed:
Update dualiso.c

fijha committed:
Dual ISO autocorrection update/bug fix

fijha committed:
Fix session state on canceling save prompt

fijha committed:
Merge branch 'ilia3101:master' into Dual-ISO

fijha committed:
Merge branch 'master' into Dual-ISO

fijha committed:
Merge branch 'ilia3101:master' into Dual-ISO

fijha committed:
Dual ISO autocorrection update/bug fix

fijha committed:
Dual ISO black delta slider step size

masc4ii commented:
@GhervasaCristian : combining the changes from this PR with the changes from @Gabboxl  makes indeed no sense. Those are two different ways of doing things in parallel. Here the algorithm is multithreaded in a "one clip export", and in the other solution best results come with many single threaded clip exports. For preview the other solution won't help. Maybe it'll be possible to switch OMP to single threaded, when using the export methode from @Gabboxl , then we could have fast preview and (more) ideal (dng) export.

fijha committed:
Dual ISO export fix
